File: Classwork/dataStructures/src/mazeAttempts/maze5/Maze.java
package mazeAttempts.maze5;

import javax.media.CannotRealizeException;
import javax.media.Manager;
import javax.media.NoPlayerException;
import javax.media.Player;
import javax.sound.sampled.Clip;
import javax.swing.*;
import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

import static java.awt.Color.*;
import static java.awt.Font.ITALIC;
import static java.awt.Font.PLAIN;
import static java.lang.System.err;
import static java.lang.System.out;
import static javax.swing.JFrame.EXIT_ON_CLOSE;

public class Maze extends JPanel implements KeyListener, MouseListener {

	JFrame frame;
	int x = 0, y = 1;
	ArrayList<String[]> board = new ArrayList<String[]>();
	ArrayList<Wall> walls = new ArrayList<Wall>();
	ArrayList<Ceiling> ceilings = new ArrayList<Ceiling>();
	ArrayList<Wall> sideWalls = new ArrayList<Wall>();
	ArrayList<Wall> rectangles = new ArrayList<Wall>();
	boolean play = true;
	int direction = 1;
	int count = 0;
	Clip clip;
	int numWalls = 3;
	boolean drop = false;
	int currentX = 500;
	int currentY = 500;
	int bcX, bcY = 0;
	int pressed = 0;
	int xLength = 350;
	int yLength = 30;
	int r1 = 255, b2 = 255;
	int g1 = 0, b1 = 0, r2 = 0, g2 = 0;
	boolean on = false;
	int moves = 0;

	public Maze() {
		setBoard();
		frame = new JFrame();
		frame.add(this);
		frame.setDefaultCloseOperation(EXIT_ON_CLOSE);
		Player player = null;
		try {
			player = Manager.createRealizedPlayer(new File("Classwork/dataStructures/src/mazeAttempts/Background-Music.mp3").toURI().toURL());
		} catch (IOException e) {
			e.printStackTrace();
		} catch (NoPlayerException e) {
			e.printStackTrace();
		} catch (CannotRealizeException e) {
			e.printStackTrace();
		}
		player.start();
		frame.setSize(1500, 800);
		frame.setVisible(true);
		frame.addKeyListener(this);
		frame.addMouseListener(this);

		this.addMouseListener(this);
	}

	public static void main(String... args) {
		new Maze();
	}

	public void paintComponent(Graphics g) {
		super.paintComponent(g);
		Graphics2D gpart2 = (Graphics2D) g;
		g.setColor(BLACK);    //this will set the background color
		g.fillRect(0, 0, 1500, 800);
		//drawBoard here!
		//setBoard();
		g.setColor(WHITE);

		//2D STUFF
		for (int i = 0; i < board.size(); i++) {
			for (int j = 0; j < board.get(i).length; j++) {
				if (board.get(i)[j].equals("#"))
					g.fillRect((j * 8) + 1, (i * 8) + 250, 5, 5);
			}
		}

		g.setColor(YELLOW);

		int lastY = board.get(0).length - 1;
		int lastX = board.size() - 1;

		g.fillRect((lastY * 10) + 180, (15 * 10) + 180, 5, 5);

		g.setColor(MAGENTA);
		g.fillOval((x * 8) + 1, (y * 8) + 249, 7, 7);

		//3D Stuff
		for (Ceiling c : ceilings) {
			//g.setColor(Color.GRAY);
			gpart2.setPaint(c.getPaint());
			g.fillPolygon(c.getPolygon());

		}
		for (Wall w : rectangles) {
			g.setColor(BLACK);
			g.fillPolygon(w.getPolygon());
		}
		for (Wall w : walls) {
			g.setColor(WHITE);
			//g2.setPaint(w.getPaint());
			g.fillPolygon(w.getPolygon());
			g.setColor(BLUE);
			g.drawPolygon(w.getPolygon());
		}
		//front
		for (Wall w : sideWalls) {
			gpart2.setPaint(w.getPaint());
			//g.setColor(new Color(67,70,76));
			g.fillPolygon(w.getPolygon());
		}
		g.setFont(new Font("ALGERIAN", ITALIC, 50));
		g.setColor(new Color(0, 80, 255));
		g.drawString("Happy Maze-ing!", 10, 100);
		g.drawString("Moves Used: " + count, 40, 600);
		gpart2.setPaint(new GradientPaint(35, 175, new Color(r1, b1, g1), 35 + xLength, 175 + yLength, new Color(r2, b2, g2)));
		g.fillRect(35, 175, xLength, yLength);
		g.setFont(new Font("Lucida Calligraphy", PLAIN, 20));
		g.setColor(YELLOW);
		if (xLength != 0)
			g.drawString("Hint: Press F to use the flashlight", 35, 150);
		else {
			g.drawString("Unfortunately the flashlight\'s", 50, 150);
			g.drawString("battery has died", 115, 180);
		}
		if (on)
			g.drawString("F Pressed", 150, 230);

		if (x == lastY && y == 15) {
			g.setFont(new Font("Stencil", PLAIN, 40));
			g.setColor(WHITE);
			g.fillRect(0, 0, 1500, 800);
			g.setColor(new Color(237, 187, 249));
			g.fillRect(185, 95, 975, 500);
			g.setColor(WHITE);
			g.fillRect(190, 100, 965, 490);
			g.setColor(new Color(49, 224, 177));
			g.drawString("Congratulations on completing the maze!", 225, 325);
			g.drawString("You used " + count + " moves", 500, 425);
		}
	}

	public void setBoard() {

		File name = new File("Classwork/dataStructures/src/mazeAttempts/Maze.txt");
		try {
			BufferedReader input = new BufferedReader(new FileReader(name));
			String text;
			while ((text = input.readLine()) != null) {
				board.add(text.split(""));
			}
		} catch (IOException io) {
			err.println("File error");
		}
		for (int r = 0; r < board.size(); r++) {
			for (int c = 0; c < board.get(r).length; c++)
				out.print(board.get(r)[c]);
			out.println();
		}
		ceilingStuff();
		rightRectangle();
		leftRectangle();
		leftSide();
		rightSide();
	}

	public void leftSide() {
		for (int i = 0; i < numWalls; i++) {
			int[] xCoordinates = {0, 50, 650, 700};
			int[] yCoordinates = {450, 550, 550, 450};

			for (int j = 0; j < xCoordinates.length; j++) {
				if (j < 2)
					xCoordinates[j] += 50 * i;
				else
					xCoordinates[j] -= 50 * i;
				yCoordinates[j] += 100 * i;
			}

			if (direction == 1) { //right
				if ((x + i) < board.get(0).length && y - 1 >= 0 && board.get(y - 1)[x + i].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			} else if (direction == 2) { //down
				if ((y + i) < board.size() && x + 1 < board.get(y + i).length && board.get(y + i)[x + 1].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			} else if (direction == 3) { //left
				if ((x - i) >= 0 && y + 1 < board.size() && board.get(y + 1)[x - i].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			} else if (direction == 4) { //up
				if ((y - i) >= 0 && x - 1 >= 0 && board.get(y - i)[x - 1].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			}
		}
	}

	public void flashlightPower() {
		xLength -= 25;
		if (b2 >= 0 && r2 <= 255) {
			b2 -= 10;
			r2 += 15;
		}
	}

	public void rightSide() {
		for (int i = 0; i < numWalls; i++) {
			int[] xCoordinates = {0, 50, 650, 700};
			int[] yCoordinates = {1300, 1200, 1200, 1300};
			for (int j = 0; j < xCoordinates.length; j++) {
				if (j < 2)
					xCoordinates[j] += 50 * i;
				else
					xCoordinates[j] -= 50 * i;
				yCoordinates[j] -= 100 * i;
			}
			if (direction == 1) { //right
				if ((x + i) < board.get(0).length && (y + 1) < board.size() && board.get(y + 1)[x + i].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			} else if (direction == 2) { //down
				if ((y + i) < board.size() && x - 1 >= 0 && board.get(y + i)[x - 1].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			} else if (direction == 3) { //left
				if ((x - i) >= 0 && (y - 1) >= 0 && board.get(y - 1)[x - i].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			} else if (direction == 4) { //up
				if ((y - i) >= 0 && x + 1 < board.get(y - i).length && board.get(y - i)[x + 1].equals("#"))
					walls.add(new Wall(xCoordinates, yCoordinates));
			}
		}
	}

	public void leftRectangle() {

		for (int i = 0; i < numWalls; i++) {
			int[] xCoordinates = {50, 50, 650, 650};
			int[] yCoordinates = {450, 550, 550, 450};

			if ((x + 1) < board.get(0).length) {
				for (int j = 0; j < xCoordinates.length; j++) {
					if (j < 2)
						xCoordinates[j] += 50 * i;
					else
						xCoordinates[j] -= 50 * i;
					yCoordinates[j] += 100 * i;
				}
				rectangles.add(new Wall(xCoordinates, yCoordinates));
			}
		}
	}

	public void rightRectangle() {
		for (int i = 0; i < numWalls; i++) {
			int[] xCoordinates = {50, 50, 650, 650};
			int[] yCoordinates = {1300, 1200, 1200, 1300};

			if ((x + 1) < board.get(0).length) {
				for (int j = 0; j < xCoordinates.length; j++) {
					if (j < 2)
						xCoordinates[j] += 50 * i;
					else
						xCoordinates[j] -= 50 * i;
					yCoordinates[j] -= 100 * i;
				}
				rectangles.add(new Wall(xCoordinates, yCoordinates));
			}
		}
	}

	public void ceilingStuff() {
		for (int i = 0; i < 13; i++) {
			int[] clX = {450, 450, 1300, 1300};
			int[] clY = {0, 150, 150, 0};

			for (int j = 0; j < clX.length; j++) {
				clY[j] += 50 * i;
			}
			ceilings.add(new Ceiling(clX, clY));
		}
	}

	public void frontWall() {
		int[] xStuff = {0, 0, 0, 0};
		int[] yStuff = {0, 0, 0, 0};
		sideWalls = new ArrayList<Wall>();

		for (int i = numWalls; i >= 0; i--) {
			int[] vert = {0, 0, 700, 700};
			int[] horiz = {450, 1300, 1300, 450};
			for (int j = 0; j < 4; j++) {
				if (j < 2)
					vert[j] += 50 * i;
				else
					vert[j] -= 50 * i;
			}
			horiz[0] += 100 * i;
			horiz[1] -= 100 * i;
			horiz[2] -= 100 * i;
			horiz[3] += 100 * i;

			if (direction == 1) { //right
				if (x + i < board.get(0).length && board.get(y)[x + i].equals("#"))
					sideWalls.add(new Wall(vert, horiz));
			}
			if (direction == 2) { //down
				if (y + i < board.size() && board.get(y + i)[x].equals("#"))
					sideWalls.add(new Wall(vert, horiz));
			}
			if (direction == 3) { //left
				if (x - i >= 0 && board.get(y)[x - i].equals("#"))
					sideWalls.add(new Wall(vert, horiz));
			}
			if (direction == 4) { //up
				if (y - i >= 0 && board.get(y - i)[x].equals("#"))
					sideWalls.add(new Wall(vert, horiz));
			}
		}
	}

	public void keyPressed(KeyEvent e) {
		if (play) {
			if (e.getKeyCode() == KeyEvent.VK_UP) {
				if (direction == 1) { //right
					if ((x + 1) < board.get(0).length && !(board.get(y)[x + 1].equals("#"))) {
						x++;
						count++;
					}
				}
				if (direction == 2) { //down
					if ((y + 1) < board.size() && !(board.get(y + 1)[x].equals("#"))) {
						y++;
						count++;
					}
				}
				if (direction == 3) { //left
					if ((x - 1) < board.get(0).length && !(board.get(y)[x - 1].equals("#"))) {
						x--;
						count++;
					}
				}
				if (direction == 4) { //up
					if ((y - 1) < board.size() && !(board.get(y - 1)[x].equals("#"))) {
						y--;
						count++;
					}
				}
				moves++;
			}
			if (xLength > 0) {
				if (e.getKeyCode() == KeyEvent.VK_F) {
					if (numWalls == 3) {
						numWalls = 4;
						on = true;
					} else {
						numWalls = 3;
						on = false;
					}
				}
			} else {
				numWalls = 3;
				on = false;
			}
			if (e.getKeyCode() == KeyEvent.VK_RIGHT) {
				direction++;
				if (direction == 5)
					direction = 1;
			}
			if (e.getKeyCode() == KeyEvent.VK_LEFT) {
				direction--;
				if (direction == 0)
					direction = 4;
			}
			if (moves == 3) {
				moves = 0;
				if (on)
					flashlightPower();
			}
		}

		walls = new ArrayList<Wall>();
		rectangles = new ArrayList<Wall>();
		ceilingStuff();
		rightRectangle();
		leftRectangle();
		rightSide();
		leftSide();
		frontWall();

		repaint();
	}

	public void keyReleased(KeyEvent e) {
	}

	public void keyTyped(KeyEvent e) {
	}

	public void mouseClicked(MouseEvent e) {
		bcX = e.getX();
		bcY = e.getY();
		out.println(bcX + "," + bcY);
	}

	public void mousePressed(MouseEvent e) {
	}

	public void mouseReleased(MouseEvent e) {
	}

	public void mouseEntered(MouseEvent e) {
	}

	public void mouseExited(MouseEvent e) {
	}

	//INNER CLASS FOR CEILING PRINTING
	public class Ceiling {
		int xStuff[];
		int yStuff[];

		public Ceiling(int[] x, int[] y) {
			xStuff = x;
			yStuff = y;

		}

		public Polygon getPolygon() {
			return new Polygon(xStuff, yStuff, xStuff.length);
		}

		public GradientPaint getPaint() {
			return new GradientPaint(xStuff[0], yStuff[0], new Color(47, 79, 79), xStuff[0], yStuff[2], new Color(192, 192, 192));
		}
	}

	//INNER CLASS FOR WALL PRINTING
	public class Wall {
		int xStuff[];
		int yStuff[];

		public Wall(int[] x, int[] y) {
			xStuff = x;
			yStuff = y;

		}

		public Polygon getPolygon() {
			return new Polygon(yStuff, xStuff, xStuff.length);
		}

		public GradientPaint getPaint() {
			return new GradientPaint(xStuff[0], yStuff[0], new Color(187, 184, 193), xStuff[2], yStuff[2], new Color(27, 16, 48));
		}
	}
}

File: Classwork/dataStructures/src/maze/Maze.kt
package maze

import java.awt.Color.*
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.FileReader
import javax.swing.JFrame
import javax.swing.JPanel

class Maze : JPanel(), KeyListener, MouseListener {
	var frame: JFrame
	lateinit var maze: Array<Array<String?>>
	var explorer: Explorer? = null
	var isPainted = false
	var alreadyRead = false
	var lineArr = arrayOfNulls<String>(33)
	var k = 0
	var line: String? = null
	var lineCount = 0
	var draw3D = false
	var shrink = 50
	lateinit var walls: ArrayList<Wall>
	public override fun paintComponent(g: Graphics) {
		super.paintComponent(g)
		val g2 = g as Graphics2D
		g.setColor(BLACK)
		g.fillRect(0, 0, 1600, 1000)
		k = 0
		if (draw3D == false) {
			maze = Array(34) { arrayOfNulls<String>(34) }
//			maze = new String[lineArr.length][lineArr[0].split("").length];
			for (i in lineArr.indices) lineArr[0]?.split("".toRegex())?.toTypedArray()?.indices?.forEach {
				maze[i][it] = lineArr[i]?.split("".toRegex())?.toTypedArray()?.get(it)
				if (maze[i][it] == "#") g.setColor(WHITE) else if (maze[i][it] == "1" && isPainted == false) {
					explorer = Explorer(RED, Location(it, i), 15, 0)
					g.setColor(explorer?.color)
					isPainted = true
				} else if (maze[i][it] == "E") g.setColor(GREEN) else g.setColor(BLACK)
				g.fillRect(it * 15, i * 15, 15, 15)
			}
			if (isPainted == true) {
				g.setColor(explorer?.color)
				g2.fill(explorer?.rectangle)
			}
		} else if (draw3D) walls.indices.forEach {
			g2.color = WHITE
			g2.fillPolygon(walls[it].polygon)
			g2.color = RED
			g2.drawPolygon(walls[it].polygon)
		}
	}

	fun createWalls() {
		walls = ArrayList()
		for (n in 0..4) {
			walls.add(getCeil(n))
			walls.add(getFloor(n))
			//walls.add(getLeft(n));
			//walls.add(getRight(n));
		}
		val explorerX = explorer?.location?.x
		val explorerY = explorer?.location?.y
		val explorerDirection = explorer?.dir
		when (explorerDirection) {
			0 -> {
				var n = 0
				while (n < 5) {
					try {
						//if(maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[explorerY!! + 1][explorerX!! + n] == "#") walls.add(getRight(n))
						if (maze[explorerY - 1][explorerX + n] == "#") walls.add(getLeft(n))
					} catch (e: ArrayIndexOutOfBoundsException) {
						e.printStackTrace()
					}
					n++
				}
			}
			1 -> {
				var n = 0
				while (n < 5) {
					try {
						//if(maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[explorerY!! + n][explorerX!! - 1] == "#") walls.add(getRight(n))
						if (maze[explorerY + n][explorerX + 1] == "#") walls.add(getLeft(n))
					} catch (e: ArrayIndexOutOfBoundsException) {
						e.printStackTrace()
					}
					n++
				}
			}
			2 -> {
				var n = 0
				while (n < 5) {
					try {
						//if(maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[explorerY!! + 1][explorerX!! - n] == "#") walls.add(getRight(n))
						if (maze[explorerY - 1][explorerX - n] == "#") walls.add(getLeft(n))
					} catch (e: ArrayIndexOutOfBoundsException) {
						e.printStackTrace()
					}
					n++
				}
			}
			3 -> {
				var n = 0
				while (n < 5) {
					try {
						//if(maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[explorerY!! - n][explorerX!! + 1] == "#") walls.add(getRight(n))
						if (maze[explorerY - n][explorerX - 1] == "#") walls.add(getLeft(n))
					} catch (e: ArrayIndexOutOfBoundsException) {
						e.printStackTrace()
					}
					n++
				}
			}
			-1 -> {
				var n = 0
				while (n < 5) {
					try {
						//if(maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[explorerY!! - n][explorerX!! + 1] == "#") walls.add(getRight(n))
						if (maze[explorerY - n][explorerX - 1] == "#") walls.add(getLeft(n))
					} catch (e: ArrayIndexOutOfBoundsException) {
						e.printStackTrace()
					}
					n++
				}
			}
			-2 -> {
				var n = 0
				while (n < 5) {
					try {
						//if(maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[explorerY!! + 1][explorerX!! - n] == "#") walls.add(getRight(n))
						if (maze[explorerY - 1][explorerX - n] == "#") walls.add(getLeft(n))
					} catch (e: ArrayIndexOutOfBoundsException) {
						e.printStackTrace()
					}
					n++
				}
			}
			-3 -> {
				var n = 0
				while (n < 5) {
					try {
						//if(maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[explorerY!! + n][explorerX!! - 1] == "#") walls.add(getRight(n))
						if (maze[explorerY + n][explorerX + 1] == "#") walls.add(getLeft(n))
					} catch (e: ArrayIndexOutOfBoundsException) {
						e.printStackTrace()
					}
					n++
				}
			}
		}
	}

	fun getFloor(n: Int) = Wall(
		intArrayOf(700 - n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink),
		intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink),
		"Floor"
	)


	fun getCeil(n: Int) = Wall(
		intArrayOf(100 + n * shrink, 150 + n * shrink, 150 + n * shrink, 100 + n * shrink),
		intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink),
		"Ceil"
	)

	fun getLeft(n: Int) = Wall(
		intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink),
		intArrayOf(100 + n * shrink, 150 + n * shrink, 150 + n * shrink, 100 + n * shrink),
		"Left"
	)

	fun getRight(n: Int) = Wall(
		intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink),
		intArrayOf(700 - n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink),
		"Right"
	)

	fun getFront(n: Int) = Wall(
		intArrayOf(100 + n * shrink, 100 + n * shrink, 700 - n * shrink, 700 - n * shrink),
		intArrayOf(100 + n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink),
		"Ceil"
	)

	override fun keyPressed(e: KeyEvent) {
		if (e.keyCode == 32) draw3D = !draw3D
		if (draw3D) createWalls()
		explorer?.move(e.keyCode, maze, draw3D)
		repaint()
	}

	override fun keyReleased(e: KeyEvent) {}
	override fun keyTyped(e: KeyEvent) {}
	override fun mouseClicked(e: MouseEvent) {}
	override fun mousePressed(e: MouseEvent) {}
	override fun mouseReleased(e: MouseEvent) {}
	override fun mouseEntered(e: MouseEvent) {}
	override fun mouseExited(e: MouseEvent) {}

	companion object {
		@JvmStatic
		fun main(args: Array<String>) {
			Maze()
		}
	}

	init {
		frame = JFrame()
		frame.add(this)
		frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
		frame.setSize(1600, 1000)
		frame.isVisible = true
		frame.addKeyListener(this)
		frame.addMouseListener(this)
		try {
			val reader = BufferedReader(FileReader("Classwork/dataStructures/src/maze/Maze.txt"))
			while (reader.readLine().also { line = it } != null) {
				lineArr[k] = line
				lineCount++
				k++
			}
		} catch (e: FileNotFoundException) {
			e.printStackTrace()
		}
	}
}
/*
import java.awt.Color.*
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.event.KeyEvent
import java.awt.event.KeyEvent.*
import java.awt.event.KeyListener
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import java.io.File
import java.io.IOException
import java.lang.System.err
import java.util.*
import javax.swing.JFrame
import javax.swing.JFrame.EXIT_ON_CLOSE
import javax.swing.JPanel

class Maze : JPanel(), KeyListener, MouseListener {
	var frame: JFrame
	var list = ArrayList<String>()
	var walls2D = ArrayList<Location>()
	var walls = ArrayList<Wall>()
	var draw3D = false
	var explorerLocationIndex = 0
	var shrink = 15
	var explorer = Explorer(Location(1, 2), 0, 13, white)
	fun setBoard() {
		try {
			Scanner(File("Classwork/dataStructures/src/maze/Maze.txt")).use {
				var txt: String
				do {
					txt = it.nextLine()
					list.add(txt)
					for (i in 0 until txt.length) if (txt[i] == '1') explorerLocationIndex = txt.indexOf('1')
					repaint()
				} while (it.hasNextLine())
			}
		} catch (e: IOException) {
			err.print("File not found!")
		}
	}

	public override fun paintComponent(g: Graphics) {
		super.paintComponent(g)
		val g2 = g as Graphics2D
		repaint()
		if (!draw3D) {
			for (i in list.indices) (0 until list[i].length).forEach { j ->
				with(list[i].split("")*/
/*.toTypedArray()*//*
) {
					if (this[j] == "#") {
						g.setColor(black)
						g.fillRect(j * 13, i * 13, 13, 13)
						walls2D.add(explorer.location)
					} else if (this[j] == " ") {
						g.setColor(YELLOW)
						g.fillRect(j * 13, i * 13, 13, 13)
					} else if (this[j] == "1") {
						g.setColor(YELLOW)
						g.fillRect(j * 13, i * 13, 13, 13)
					} else if (this[j] == "E") {
						g.setColor(BLUE)
						g.fillRect(j * 13, i * 13, 13, 13)
					}
				}
			}
			g.setColor(explorer.color)
			g.fillRect(explorer.location.x * 13, explorer.location.y * 13, 13, 13)
		} else {
			for (w in walls) {
				g2.color = WHITE
				g2.fillPolygon(w.polygon)
				g2.color = BLACK
				g2.drawPolygon(w.polygon)
			}
		}
	}

	fun createWall() {
		walls = ArrayList()
		for (n in 0..4) {
			walls.add(getCeiling(n))
			walls.add(getFloor(n))
		}
	}

	fun getCeiling(n: Int) = Wall(
		intArrayOf(100 + n * shrink, 150 + n * shrink, 150 + n * shrink, 100 + n * shrink),
		intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink),
		"Ceil"
	)

	fun getFloor(n: Int) = Wall(
		intArrayOf(700 - n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink),
		intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink),
		"Floor"
	)

	override fun keyPressed(e: KeyEvent) {
		if (e.keyCode == VK_SPACE) draw3D = !draw3D
		if (draw3D) createWall()
		if (e.keyCode == VK_UP && list[explorer.location.y - 1][explorer.location.x] != '#') {
			println("up")
			explorer.move(0)
			println(explorer.location.y.toString() + " " + explorer.location.x)
			repaint()
		}
		if (e.keyCode == VK_LEFT && list[explorer.location.y][explorer.location.x - 1] != '#') {
			println("left")
			explorer.move(1)
			println(explorer.location.y.toString() + " " + explorer.location.x)
			repaint()
		}
		if (e.keyCode == VK_RIGHT && list[explorer.location.y][explorer.location.x + 1] != '#') {
			println("right")
			explorer.move(2)
			println(explorer.location.y.toString() + " " + explorer.location.x)
			repaint()
		}
		if (e.keyCode == VK_DOWN && list[explorer.location.y + 1][explorer.location.x] != '#') {
			println("down")
			explorer.move(3)
			println(explorer.location.y.toString() + " " + explorer.location.x)
			repaint()
		}
		if (explorer.location.y == 30 && explorer.location.x == 72) println("You win!")
	}

	override fun keyReleased(e: KeyEvent) {}
	override fun keyTyped(e: KeyEvent) {}
	override fun mouseClicked(e: MouseEvent) {}
	override fun mousePressed(e: MouseEvent) {}
	override fun mouseReleased(e: MouseEvent) {}
	override fun mouseEntered(e: MouseEvent) {}
	override fun mouseExited(e: MouseEvent) {}

	companion object {
		@JvmStatic
		fun main(args: Array<String>) {
			Maze().setBoard()
		}
	}

	init {
		frame = JFrame().apply {
			add(this@Maze)
			defaultCloseOperation = EXIT_ON_CLOSE
			setSize(1000, 600)
			isVisible = true
			addKeyListener(this@Maze)
			addMouseListener(this@Maze)
		}
	}
}*/
